### PR TITLE
Added Arquillian recorder integration with Jira, Github, Redmine.

### PR DESCRIPTION
**Pull request by dipak-pawar**
This change is [Reviewable](https://reviewable.io/reviews/arquillian/arquillian-governor/37)

### CODE REVIEW

**Comment by dipak-pawar** [comment]
@bartoszmajsak, We can add some functional tests for this new functionality

File: github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.TreeSet;
+
+/**
+ * @author <a href="mailto:dpawar@redhat.com">Dipak Pawar</a>
+ */
+
+public class GitHubGovernorRecorder {
+
+    @Inject
+    private Instance<GitHubGovernorConfiguration> gitHub;
```
Comment (smiklosovic) [inline github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java]: githubGovernorConfiguration

File: github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+                    uniqDetectables.add(detectable.getSimpleName());
+                }
+
+                String detectablesName = "";
+                for (String detectable : uniqDetectables
+                        ) {
+                    detectablesName += (detectablesName == "" ? "" : ",") + detectable;
+                }
+
+                TableEntry jiraDetector = new TableEntry();
```
Comment (smiklosovic) [inline github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java]: jiraDetector in GitHub? I believe this is copy paste mistake

File: github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                Class<? extends Detectable>[] detectables = gitHubValue.detector().value();
+                TreeSet<String> uniqDetectables = new TreeSet<String>();
+
+                for (Class<? extends Detectable> detectable : detectables) {
+                    uniqDetectables.add(detectable.getSimpleName());
+                }
+
+                String detectablesName = "";
+                for (String detectable : uniqDetectables
+                        ) {
```
Comment (smiklosovic) [inline github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java]: wrong formatting

File: github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                );
+                jiraDetector.getTableBody().addRow(row);
+
+                propertyReportEvent.fire(new PropertyReportEvent(new KeyValueEntry("GitHub URL", gitHubIssueURL)));
+                propertyReportEvent.fire(new PropertyReportEvent(jiraDetector));
+            }
+        }
+    }
+
+    public String contructGitHubIssueURL(GitHubGovernorConfiguration configuration, GitHub gitHub){
```
Comment (smiklosovic) [inline github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java]: why public and not private?

File: redmine/src/main/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorder.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+        final Method testMethod = event.getTestMethod();
+        final TestClass testClass = event.getTestClass();
+        final String redmineServerURL = redmineGovernorConfigurationInstance.get().getServer();
+
+        final Redmine redmineValue = getRedmineValue(testMethod, testClass);
+        if (redmineValue != null) {
+            final String issueURL = constructRedmineIssueURL(redmineServerURL, redmineValue.value());
+
+            final TableEntry jiraDetector = new TableEntry();
```
Comment (bartoszmajsak) [inline redmine/src/main/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorder.java]: Why is it named `jira`? Copy-paste?

File: jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        for (String cellContent: cellContents) {
+            row.addCell(new TableCellEntry(cellContent));
+        }
+        return row;
+    }
+
+    @Jira(value = "ARQ-234",  detector = @Detector(value = OS.Windows.class))
+    private static final class FakeTestClass {
+        @Test
+        @Jira(value = "ARQ-123", detector = @Detector(value = OS.Unix.class), force = true)
```
Comment (bartoszmajsak) [inline jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java]: Shouldn't the annotation be all uppercase in this case - `@JIRA`?

File: redmine/src/test/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorderTest.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        assertTrue(detectors.equals(getTableEntry("RedmineOptions", tableOptions)));
+    }
+
+    @Test
+    public void sholdReportRedmineGovernorParamsForClass(){
+        try {
+            After after = new After(new FakeTestClass(), FakeTestClass.class.getMethod("someTestMethod"));
+            redmineGovernorRecorder.redmineReportEntries(after);
+        } catch (NoSuchMethodException e) {
+            e.printStackTrace();
```
Comment (bartoszmajsak) [inline redmine/src/test/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorderTest.java]: Why are we doing it? Shouldn't the test fail in such a case?

File: jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    @Test
+    public void shouldReportJiraGovernorParamsForMethod() {
+
+        try {
+            After after = new After(FakeTestClass.class, FakeTestClass.class.getMethod("dummyTest"));
+              jiraGovernorRecorder.jiraReportEntries(after);
+        } catch (NoSuchMethodException e) {
+            e.printStackTrace();
```
Comment (bartoszmajsak) [inline jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java]: Why are we doing it? Shouldn't the test fail in such a case?

File: jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class JiraGovernorRecorderTest {
```
Comment (bartoszmajsak) [inline jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java]: I think these tests need some clean up. They assert too much and have quite mixed flow which is hard to follow. Can you try to distill them to some concrete features we would like to verify and have simple flow of

  * setup initial context for the test
  * exercise components under tests
  * verify state or behaviour using assertions or mock verifications at the end?

I would also recommend to use [assertj](http://joel-costigliola.github.io/assertj/) for fluent and readable assertions.

File: redmine/src/main/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorder.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+import org.jboss.arquillian.core.api.Instance;
+import org.jboss.arquillian.core.api.annotation.Inject;
+import org.jboss.arquillian.core.api.annotation.Observes;
+import org.jboss.arquillian.test.spi.TestClass;
+import org.jboss.arquillian.test.spi.event.suite.After;
+
+import java.lang.reflect.Method;
+
+/**
+ * @author <a href="mailto:dpawar@redhat.com">Dipak Pawar</a>
```
Comment (bartoszmajsak) [inline redmine/src/main/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorder.java]: I'm not sure if we should have authors in the code, as it's collective code ownership afterwards. But I guess it won't harm to add Apache license headers like in the [Arquillian Core](https://github.com/bartoszmajsak/arquillian-core/blob/b47c8779fe2e56e0714bf840e178d4d146e2da88/junit/core/src/test/java/org/jboss/arquillian/junit/rules/AbstractRuleStatementEnrichment.java#L1-L16). WDYT @lordofthejars @smiklosovic ?

Comment (dipak-pawar) [inline redmine/src/main/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorder.java]: @bartoszmajsak, I am not sure about authors. but for Apache license headers, maven checkstyle plugin is looking for Apache licence headers. If it doesn't found it's giving error for this project.

File: jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java (hunk 6 above)
Comment (dipak-pawar) [inline jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java]: nopes annotation is defined like @Jira.

File: redmine/src/main/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorder.java (hunk 5 above)
Comment (dipak-pawar) [inline redmine/src/main/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorder.java]: Yes copy-paste

File: github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+                }
+
+                String detectablesName = "";
+                for (String detectable : uniqDetectables
+                        ) {
+                    detectablesName += (detectablesName == "" ? "" : ",") + detectable;
+                }
+
+                TableEntry jiraDetector = new TableEntry();
+                jiraDetector.setTableName("GitHubOptions");
```
Comment (dipak-pawar) [inline github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java]: need to change for gitHub. it's due to copy-paste

File: jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java (hunk 8 above)
Comment (dipak-pawar) [inline jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java]: We are testing logic in jiraReportEntries method, Nopes test shouldn't fail in this case.

File: jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java (hunk 6 above)
Comment (bartoszmajsak) [inline jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java]: Sure, I was just wondering if we shouldn't change it to uppercase.

File: redmine/src/main/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorder.java (hunk 10 above)
Comment (bartoszmajsak) [inline redmine/src/main/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorder.java]: So we have to bring them anyway :) IDEA has a way to automate it.

File: jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java (hunk 8 above)
Comment (bartoszmajsak) [inline jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java]: What's the reason for this `try-catch`? It makes this already big test more noisy.

Comment (dipak-pawar) [inline jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java]: It's just to make sure we are catching exception if named method is not found `<FakeTestClass.class.getMethod("dummyTest")` for this code. I'll try to make it short & less noisy. 

One more thing assertj looks very good if you compare it with junit default assertions.

Comment (bartoszmajsak) [inline jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java]: > FakeTestClass.class.getMethod("dummyTest")

but in this case it's implemented so this branch will never be reached. And in real usage it probably will never occur at all.

Comment (dipak-pawar) [inline jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java]: I agree, in real situation this branch will never be reached. but we are using getMethod which is throwing NoSuchMethodException, We should caught it or declare it to be thrown to avoid compilation problem.

In this case I should use throws instead of try catch.

Comment (bartoszmajsak) [inline jira/src/test/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorderTest.java]: Then let's just wrap it in the method and use it the tests instead.  Checked exceptions... always a problem :D

File: github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+            gitHubDetector.getTableBody().addRow(row);
+
+            propertyReportEvent.fire(new PropertyReportEvent(new KeyValueEntry("GitHub URL", gitHubIssueURL)));
+            propertyReportEvent.fire(new PropertyReportEvent(gitHubDetector));
+        }
+    }
+
+    private String contructGitHubIssueURL(GitHubGovernorConfiguration configuration, GitHub gitHub) {
+        String url = configuration.getDefaultGithubURL();
+        if (configuration.getRepositoryUser() != "") {
```
Comment (bartoszmajsak) [inline github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java]: we can also use `.isEmpty()` assuming we are on JDK 6+

File: github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+            url += ("/" + configuration.getRepositoryUser());
```
Comment (bartoszmajsak) [inline github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java]: Why in `()`?

File: jira/src/main/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorder.java [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+
+        final TestClass testClass = event.getTestClass();
+        final String jiraServer = jiraGovernorConfigurationInstance.get().getServer();
+
+        final Jira jira = getJiraValue(method, testClass);
+        if (jira != null) {
+            final String issueURL = constructJiraIssueURL(jiraServer, jira.value());
+
+            final Class<? extends Detectable>[] detectables = jira.detector().value();
+            final TreeSet<String> uniqDetectables = new TreeSet<String>();
```
Comment (bartoszmajsak) [inline jira/src/main/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorder.java]: We can use `Set` instead of concrete type.

File: pom.xml [hunk 15]
```diff
@@ -160,6 +166,25 @@
...
+                <artifactId>arquillian-recorder-reporter-spi</artifactId>
+                <version>${version.arquillian.recorder}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.arquillian.extension</groupId>
+                <artifactId>arquillian-recorder-reporter-api</artifactId>
+                <version>${version.arquillian.recorder}</version>
+            </dependency>
+
+            <!-- assertj dependency for assetion -->
```
Comment (bartoszmajsak) [inline pom.xml]: No need for this comment.

File: redmine/pom.xml [hunk 16]
```diff
@@ -102,6 +116,13 @@
             <groupId>org.hamcrest</groupId>
             <artifactId>hamcrest-junit</artifactId>
         </dependency>
+
+        <!--AssertJ for assertion -->
```
Comment (bartoszmajsak) [inline redmine/pom.xml]: No need for this comment.

File: redmine/src/test/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorderTest.java [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+            method = FakeTestClass.class.getMethod(methodName);
+        } catch (NoSuchMethodException e) {
+            e.printStackTrace();
+        }
+        return method;
+    }
+
+    @Redmine(value = "1",  force = true)
+    private static final class FakeTestClass {
+        @Test
```
Comment (bartoszmajsak) [inline redmine/src/test/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorderTest.java]: Just wondering - are those `@Test`s also run by JUnit?

File: README.adoc [hunk 18]
```diff
@@ -188,6 +188,28 @@ public class IssuesWithConditionalTestCases
 }
 ----
 
+==== Arquillian Recorder Integration with Arquillian Jira Governor Extension
+
+How to get Jira issues,force values, detector & it's strategy for environment  in arquillian reports?
```
Comment (bartoszmajsak) [inline README.adoc]: Small formatting issues.

File: README.adoc [hunk 19]
```diff
@@ -188,6 +188,28 @@ public class IssuesWithConditionalTestCases
...
+
+In order to see all issues & detectors in arquillian reports, you have to put arquillian-recorder-reporter dependency in your Arquillian maven test.
```
Comment (bartoszmajsak) [inline README.adoc]: it's not `maven` test I believe.

File: README.adoc [hunk 20]
```diff
@@ -329,6 +351,28 @@ public class IssuesWithConditionalTestCases
 }
 ----
 
+==== Arquillian Recorder Integration with Arquillian GitHub Governor Extension
+
+How to get GitHub issues,force values, detector & it's strategy for environment  in arquillian reports?
```
Comment (bartoszmajsak) [inline README.adoc]: As above - formatting.

File: README.adoc [hunk 21]
```diff
@@ -329,6 +351,28 @@ public class IssuesWithConditionalTestCases
...
+ 
+In order to see all issues & detectors in arquillian reports, you have to put arquillian-recorder-reporter dependency in your Arquillian maven test.
```
Comment (bartoszmajsak) [inline README.adoc]: As above - not really mvn test.

File: README.adoc [hunk 22]
```diff
@@ -559,6 +603,28 @@ added into the tree of Arquillian Report output file in every case.
 
 |===
 
+==== Arquillian Recorder Integration with Arquillian Redmine Governor Extension
+
+How to get Redmine issues,force value in arquillian reports?
+
+In order to see all issues & force value in arquillian reports, you have to put arquillian-recorder-reporter dependency in your Arquillian maven test.
+
```
Comment (bartoszmajsak) [inline README.adoc]: As above.

File: github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java (hunk 2 above)
Comment (dipak-pawar) [inline github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java]: yes.

File: README.adoc [hunk 23]
```diff
@@ -559,6 +603,28 @@ added into the tree of Arquillian Report output file in every case.
...
+    <groupId>org.arquillian.extension</groupId>
+    <artifactId>arquillian-recorder-reporter-impl</artifactId>
+    <version>${version.arquillian.recorder}</version>
+    <scope>test</scope>
+</dependency>
+----
+
+Configure arquillian-recorder-reporter as your requirement. Follow steps from here - https://github.com/arquillian/arquillian-recorder/wiki/Reporter
+
+You will get Redmine's all issues & force values i.e. true or false information located in target/arquillian-report.xml report.
```
Comment (bartoszmajsak) [inline README.adoc]: What does it mean?

> true or false information

File: github/pom.xml [hunk 24]
```diff
@@ -95,6 +95,27 @@
...
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.arquillian.extension</groupId>
+            <artifactId>arquillian-recorder-reporter-api</artifactId>
+            <version>${version.arquillian.recorder}</version>
+            <scope>provided</scope>
+        </dependency>
+
+        <!--AssertJ for assertion -->
```
Comment (bartoszmajsak) [inline github/pom.xml]: No need for this comment.

File: github/pom.xml [hunk 25]
```diff
@@ -95,6 +95,27 @@
...
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <version>${version.assertj}</version>
```
Comment (bartoszmajsak) [inline github/pom.xml]: I believe we can clean up dependencies newly added in the modules.

  * move artifacts with version to parent pom `dependencyManagement` section.
  * in all modules refer w/o version.
  * use correct scope - for `assertj` it should be `test`

File: github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java (hunk 4 above)
Comment (dipak-pawar) [inline github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java]: @smiklosovic , I have updated all formatting changes.  To use Java 8 features we need to enforce user to use java8 which I am not sure we should at this point.

File: README.adoc (hunk 23 above)
Comment (dipak-pawar) [inline README.adoc]: force value has two options which you can set using `@Redmine(Value="1", force=true)`

File: github/pom.xml (hunk 24 above)
Comment (dipak-pawar) [inline github/pom.xml]: ok

File: github/pom.xml (hunk 25 above)
Comment (dipak-pawar) [inline github/pom.xml]: @bartoszmajsak Do we need to refer scope in each module if it's declared in parent module inside dependencyManagement section

Comment (dipak-pawar) [inline github/pom.xml]: hey got it. It inherets scope & version from it's parent.

File: github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java (hunk 12 above)
Comment (dipak-pawar) [inline github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java]: ok, I am updating it with .isEmpty? method. just wondering do we have any disadvantages of this approach?

File: github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java (hunk 13 above)
Comment (dipak-pawar) [inline github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java]: just to make more readable

File: jira/src/main/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorder.java (hunk 14 above)
Comment (dipak-pawar) [inline jira/src/main/java/org/arquillian/extension/governor/jira/impl/reporter/JiraGovernorRecorder.java]: You mean Reference type like this `final Set<String> uniqDetectables = new TreeSet<String>()`

File: github/pom.xml (hunk 25 above)
Comment (lordofthejars) [inline github/pom.xml]: @dipak-pawar no, if dependencyManagement sets an scope you don't need to set again

File: github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java (hunk 12 above)
Comment (lordofthejars) [inline github/src/main/java/org/arquillian/extension/governor/github/impl/reporter/GitHubGovernorRecorder.java]: just to be more readable

File: redmine/src/test/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorderTest.java (hunk 17 above)
Comment (lordofthejars) [inline redmine/src/test/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorderTest.java]: Probably not, but I think this test class is sued as helper inside the enclosing test

Comment (dipak-pawar) [inline redmine/src/test/java/org/arquillian/extension/governor/redmine/impl/reporter/RedmineGovernorRecorderTest.java]: @lordofthejars Yes this test class is used as helper inside enclosing test. I am not sure about `@Test` also run by JUnit means.

File: github/pom.xml (hunk 25 above)
Comment (dipak-pawar) [inline github/pom.xml]: Thanks

**Comment by dipak-pawar** [comment]
@lordofthejars I am done with changes suggested by you & Bartosz. Can you take a final look?
